app/presenters/publish_warning_presenter.rb

```diff
@@ -0,0 +1,73 @@
+class PublishWarningPresenter
+  attr_reader :service, :action
+
+  MESSAGE = {
+    publish: {
+      both_pages: I18n.t('publish.warning.both_pages'),
+      cya_page: I18n.t('publish.warning.cya'),
+      confirmation_page: I18n.t('publish.warning.confirmation')
+    },
+    delete: {
+      both_pages: I18n.t('pages.flow.delete_warning_both_pages'),
+      cya_page: I18n.t('pages.flow.delete_warning_cya_page'),
+      confirmation_page: I18n.t('pages.flow.delete_warning_confirmation_page')
+    }
+  }.freeze
+
+  def initialize(service, action)
+    @service = service
+    @action = action
+  end
+
+  def message
+    submitting_pages_not_present_message ||
+      confirmation_page_not_present_message ||
+      cya_page_not_present_message
+  end
+
+  def submitting_pages_not_present_message
+    if !checkanswers_in_main_flow? && !confirmation_in_main_flow?
+      MESSAGE[action][:both_pages]
+    end
+  end
+
+  def cya_page_not_present_message
+    if !checkanswers_in_main_flow? && confirmation_in_main_flow?
+      MESSAGE[action][:cya_page]
+    end
+  end
+
+  def confirmation_page_not_present_message
+    if checkanswers_in_main_flow? && !confirmation_in_main_flow?
+      MESSAGE[action][:confirmation_page]
+    end
+  end
+
+  def checkanswers_in_main_flow?
+    grid.page_uuids.include?(find_page_uuid('page.checkanswers'))
+  end
+
+  def confirmation_in_main_flow?
+    grid.page_uuids.include?(find_page_uuid('page.confirmation'))
+  end
+
+  def find_page_uuid(page_type)
+    if service_includes_page?(page_type)
+      matched_page =
+        service.pages.find do |page|
+          page.type == page_type
+        end
+      matched_page.uuid
+    end
+  end
+
+  def service_includes_page?(page)
+    service.pages
+      .map(&:type)
+      .include?(page)
+  end
```

Comment:
i _think_ with the two line changes above you might be able to remove `find_page_uuid` and `service_includes_page?`